pandas/core/groupby/generic.py

```diff
@@ -321,13 +321,18 @@ def _aggregate_multiple_funcs(self, arg):
                 obj = copy.copy(obj)
                 obj._reset_cache()
                 obj._selection = name
-            results[name] = obj.aggregate(func)
+            results[base.OutputKey(label=name, position=idx)] = obj.aggregate(func)
 
         if any(isinstance(x, DataFrame) for x in results.values()):
             # let higher level handle
-            return results
+            return {key.label: value for key, value in results.items()}
 
-        return DataFrame(results, columns=columns)
+        if not results:
+            return DataFrame()
+        output = self._wrap_aggregated_output(results)
+        if not isinstance(output, ABCDataFrame):
```

Comment:
hmm, what case hits this?

Reply:
@jreback here's one:

pandas/tests/groupby/test_categorical.py::test_groupby_agg_categorical_columns

```
    func = pd.Series.nunique
    df = pd.DataFrame(
        {
            "id": [0, 1, 2, 3, 4],
            "groups": [0, 1, 1, 2, 2],
            "value": pd.Categorical([0, 0, 0, 0, 1]),
        }
    ).set_index("id")
    result = df.groupby("groups").agg(func)
```

Reply:
hmm i think this last case should be handled in _wrap_aggregated_output, we shouldn't just force to a DataFrame, something else is going on

Reply:
Aren't there cases where we would want `_wrap_aggregated_output` to return a Series though?

For the test `pandas/tests/groupby/test_categorical.py::test_groupby_agg_categorical_columns`, when we enter `_wrap_aggregated_output`, we have:
```
>>> output
{OutputKey(label='value', position=0): array([1, 1, 2])}
>>> self
<pandas.core.groupby.generic.SeriesGroupBy object at 0x7f7f7f178730>
```

while for the test `pandas/tests/groupby/aggregate/test_aggregate.py::test_agg_ser_multi_key`, when we enter it, we have
```
>>> output
{OutputKey(label='C', position=0): array([ 0.3004091 , -1.87694605,  0.7516694 ,  1.01936873,  0.7758193 ,
       -0.60044377])}
>>> self
self
<pandas.core.groupby.generic.SeriesGroupBy object at 0x7f2963500160>
```

Any suggestions on how to distinguish when we'd want to return a DataFrame and when a Series?

Reply:
see my comment above, _wrap_aggregated_output should handle this

Reply:
Hi @jreback - thanks for your feedback. Sorry, I don't understand what you mean here - I can't remove
```python
if not isinstance(output, ABCDataFrame):
    output = output.to_frame()
```
from here because then `pandas/tests/groupby/test_categorical.py::test_groupby_agg_categorical_columns` would fail, as in that test `_wrap_aggregated_output` needs to return a Series.

Also, I make this change
```diff
-        return self._reindex_output(result)
+        reindexed_output = self._reindex_output(result)
+        if not isinstance(reindexed_output, DataFrame):
+            reindexed_output = reindexed_output.to_frame()
+        return reindexed_output
```
to  `_wrap_aggregated_output` because there are cases (e.g. `pandas/tests/groupby/test_categorical.py::test_groupby_agg_categorical_columns`) when `_wrap_aggregated_output` _should_ return a Series.